Multiapp reset for multiple specification of times

abarun22 opened the pull request:
This PR is associated with the issue: refs https://github.com/idaholab/moose/issues/22489

## Reason
The current way of resetting multi-apps does not provide options to perform the reset for multiple number of main app times. Hence the parameter "reset_time" has been converted to accept vector of 'double' values that will signify the range of times at which the multiapps specified in "reset_apps" will be reset

## Design
A simple conversion of "reset_time" from scalar to vector has been done. In addition the default action for the non-specification of "reset_apps" parameter was missing. If not specified the reset will be done for all associated apps.

## Impact
The function "MultiApp::preTransfer" is subject to minimal modifications that effect this new feature.

Commit:
Specification of multiple reset times ref:#22489

GiudGiud commented:
Pull requests may not be made to master, only to next.

I ll push some changes to this

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1223639/) on 762ac54 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22583/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22583/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c8ebf494b57f8d9c718405208a8445e3df28cff5`

GiudGiud committed:
Fix docstrings
Prevent multiple resets when several reset times were covered by a single timestep
Add check on ordering of reset times
Improve parameter handling for reset time

GiudGiud reviewed (dismissed):
This will need a test. 

Please be more careful with the rebase next time. The lines you deleted with the getAppInfo (I re-added them) were there for a good reason.

abarun22 commented:
My goodness. I thought that was shown as an old code during the rebase and so i would have done that. Sorry about that

GiudGiud committed:
Add test for multiple reset of multiapps
refs #22489

GiudGiud commented:
Finally had time to fix this up and add a test. 
My bad i had introduced a few errors when I fixed up the rebasing commit. They are fixed. now.

GiudGiud commented:
@cticenhour if you have a minute could you please review this?

cticenhour commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -253,15 +254,20 @@ MultiApp::MultiApp(const InputParameters & parameters)
+  auto sorted_times = _reset_times;
+  std::sort(sorted_times.begin(), sorted_times.end());
+  if (_reset_times.size() && _reset_times != sorted_times)
+    paramError("reset_time", "List of reset times must be sorted in increasing order");
```
It doesn't appear as though you are hitting this error case in your test - please add another to hit this for coverage.

cticenhour commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -516,22 +523,39 @@ MultiApp::fillPositions()
-  // First, see if any Apps need to be Reset
-  if (!_reset_happened && target_time + 1e-14 >= _reset_time)
+  // Get a transient executioner to get a user-set tolerance
+  Real timestep_tol = 1e-13;
```
Is there a reason why this was changed from `1e-14`?

GiudGiud replied:
default s changed for transient multiapps too. It s meant to be overriden anyway by what is set for the executioner

cticenhour commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -516,22 +523,39 @@ MultiApp::fillPositions()
+  Real timestep_tol = 1e-13;
+  if (dynamic_cast<Transient *>(_fe_problem.getMooseApp().getExecutioner()))
+    timestep_tol =
+        dynamic_cast<Transient *>(_fe_problem.getMooseApp().getExecutioner())->timestepTol();
```
Probably will need to include `Transient.h` here to do this `dynamic_cast`. Seeing a failure here: https://civet.inl.gov/job/1280286/

GiudGiud replied:
ah yes

cticenhour reviewed (changes requested):
You've still got some modules failures (see CIVET results). Will give a pre-emptive review anyway since this is small and I noticed a few things.

GiudGiud committed:
Address Casey's review:
- add test for sorting
- add header to avoid unity build failure

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1280436/) on 62b69ea wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22583/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1280449/) on 62b69ea wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>f4d49d</th>
      <th colspan=4>#22583 62b69e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.52%</td>
      <td>84.52%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>81534</td>
      <td>81548</td>
      <td>+14</td>
      <td>27</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14935</td>
      <td>14935</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22583/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22583/coverage/framework)


## Modules coverage

### Level set

<table>
  <thead>
    <tr>
      <th></th>
      <th>f4d49d</th>
      <th colspan=4>#22583 62b69e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.37%</td>
      <td>88.37%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>319</td>
      <td>319</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>42</td>
      <td>42</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22583/coverage_diff/level_set)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22583/coverage/level_set)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22583/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
All green now! and review should be addressed

cticenhour commented:
You have addressed my previous comment, but it looks like there is a possible related failure in Isopod (https://civet.inl.gov/job/1280434/). Though, this is the new Optimization Module right?

GiudGiud commented:
This object is duplicated in the optimization module.
@lynnmunday can it be removed from isopod?

GiudGiud commented:
Ah actually it's not duplicated, it just didnt get merged in the module.

Do you still need it?
I can patch it easily if needed

lynnmunday commented:
@GiudGiud I don't know which object is duplicated?  I don't think we need it in isopod.

cticenhour reviewed (approved):
Comments addressed and fix for isopod noted in separate issue to be addressed during optimization module transfer.

GiudGiud commented:
Thanks for the prompt review!

GiudGiud commented:
also thanks for the contribution @abarun22 !

abarun22 commented:
Thanks Guillaume
